numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -373,29 +374,24 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
-
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    static PyObject *fallback = NULL;
+    npy_cache_import("numpy.core.arrayprint", "_scalar_fallback", &fallback);
+    if (fallback == NULL) {
+        return NULL;
     }
-    return ret;
+    return PyObject_CallFunction(fallback, "O", self);
```

Comment:
Also, note the comment I added at the top of the file:

Both `scalartypes.c.src` and `arrayprint.py` implement reprs for scalars, but for different purposes. `scalartypes.c.src` has reprs for when the scalar is printed on its own, while `arrayprint.py` has reprs for when scalars are printed inside an ndarray.

This is quite apparent in datetime64, for example:
```python
>>> a = np.datetime64('2005-02-25')
>>> a
numpy.datetime64('2005-02-25')
>>> np.array([a])
array(['2005-02-25'], dtype='datetime64[D]')